### PR TITLE
SNOW-1445839: Add support for nlargest and nsmallest

### PR DESCRIPTION
**Pull request by sfc-gh-nkumar**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.
SNOW-1445839: Add support for DataFrame.nlargest
SNOW-1445840: Add support for DataFrame.nsmallest
SNOW-1445733: Add support for Series.nlargest
SNOW-1445734: Add support for Series.nsmallest
   

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

- Add support for nlargest and nsmallest methods on DataFrame and Series. Raise NotimplementedError when keep='all'
- Copy updated frontend code from modin repo.

### CODE REVIEW

File: src/snowflake/snowpark/modin/pandas/series.py [hunk 1]
```diff
@@ -1490,22 +1490,32 @@ def ne(self, other, level=None, fill_value=None, axis=0):  # noqa: PR01, RT01, D
...
         # TODO: SNOW-1063347: Modin upgrade - modin.pandas.Series functions
-        return self._default_to_pandas(pandas.Series.nlargest, n=n, keep=keep)
+        if len(self._query_compiler.columns) == 0:
+            # pandas returns empty series when requested largest/smallest from empty series
+            return self.__constructor__(data=[], dtype=float)
+        return Series(
+            query_compiler=self._query_compiler.nlargest(
+                n=n, columns=self.name, keep=keep
+            )
+        )
```
Comment (sfc-gh-nkumar) [inline src/snowflake/snowpark/modin/pandas/series.py:1505]: This is copied from modin repo.

File: src/snowflake/snowpark/modin/pandas/series.py [hunk 2]
```diff
@@ -1490,22 +1490,32 @@ def ne(self, other, level=None, fill_value=None, axis=0):  # noqa: PR01, RT01, D
...
         """
         # TODO: SNOW-1063347: Modin upgrade - modin.pandas.Series functions
+        if len(self._query_compiler.columns) == 0:
+            # pandas returns empty series when requested largest/smallest from empty series
+            return self.__constructor__(data=[], dtype=float)
         return self.__constructor__(
-            query_compiler=self._query_compiler.nsmallest(n=n, keep=keep)
+            query_compiler=self._query_compiler.nsmallest(
+                n=n, columns=self.name, keep=keep
+            )
```
Comment (sfc-gh-nkumar) [inline src/snowflake/snowpark/modin/pandas/series.py:1518]: This is copied from modin repo.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 3]
```diff
@@ -13948,6 +13948,127 @@ def dt_strftime(self, date_format: str) -> None:
...
+              If True return smallest n values otherwise return largest n values.
+
+        Returns:
+            SnowflakeQueryCompiler
+        """
+        if keep not in ("first", "last", "all"):
+            raise ValueError('keep must be either "first", "last" or "all"')
+        if keep == "all":
+            method_name = "nsmallest" if ascending else "nlargest"
+            ErrorMessage.not_implemented(
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:13975]: Do we have a ticket to track the implementation? One straightforward ways is probably do a join based on the last row

File: tests/integ/modin/frame/test_nsmallest.py
Comment (sfc-gh-mvashishtha) [inline tests/integ/modin/frame/test_nsmallest.py:1]: Let's consolidate the `nsmallest` and `nlargest` tests into a single file with a test class that we parametrize by `method` for `['nsmallest', 'nlargest']`.

File: tests/integ/modin/series/test_nlargest.py
Comment (sfc-gh-mvashishtha) [inline tests/integ/modin/series/test_nlargest.py:1]: ditto about consolidating the nlargest and nsmallest tests

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 4]
```diff
@@ -13948,6 +13948,127 @@ def dt_strftime(self, date_format: str) -> None:
...
+            internal_frame.get_snowflake_quoted_identifiers_group_by_pandas_labels(
+                columns, include_index=False
+            )
+        )
+        ordering_columns = []
+        for label, identifiers in zip(columns, matched_identifiers):
+            if len(identifiers) == 0:
+                raise KeyError(label)
+            if len(identifiers) > 1:
+                raise ValueError(f"The column label '{label} is not unique.")
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:14009]: do we have a test for this case?

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 5]
```diff
@@ -13948,6 +13948,127 @@ def dt_strftime(self, date_format: str) -> None:
...
+        if keep not in ("first", "last", "all"):
+            raise ValueError('keep must be either "first", "last" or "all"')
+        if keep == "all":
+            method_name = "nsmallest" if ascending else "nlargest"
+            ErrorMessage.not_implemented(
+                f"Snowpark pandas method '{method_name}' doesn't yet support parameter keep='all'"
+            )
+        # Special case handling for unnamed series. 'columns' passed from frontend layer
+        # will be None, replace it with MODIN_UNNAMED_SERIES_LABEL.
+        if columns is None:
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:13980]: So if column is none, it means we want to get all columns, regardless it is series or dataframe, right? if it is dataframe, we can put columns = None even with dataframe, right?

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py [hunk 6]
```diff
@@ -1892,11 +1892,222 @@ def ne():
     def nlargest():
         """
         Return the first `n` rows ordered by `columns` in descending order.
+
+        Return the first `n` rows with the largest values in `columns`, in
+        descending order. The columns that are not specified are returned as
+        well, but not used for ordering.
+
+        This method is equivalent to
+        ``df.sort_values(columns, ascending=False).head(n)``, but more
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:1901]: is it more performant? I would expect the SQL to be about the same for both.

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py [hunk 7]
```diff
@@ -1892,11 +1892,222 @@ def ne():
...
     def nsmallest():
         """
         Return the first `n` rows ordered by `columns` in ascending order.
+
+        Return the first `n` rows with the smallest values in `columns`, in
+        ascending order. The columns that are not specified are returned as
+        well, but not used for ordering.
+
+        This method is equivalent to
+        ``df.sort_values(columns, ascending=True).head(n)``, but more
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:2013]: ditto about performance

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py [hunk 8]
```diff
@@ -1892,11 +1892,222 @@ def ne():
...
+
+        When using ``keep='last'``, ties are resolved in reverse order:
+
+        >>> df.nsmallest(3, 'population', keep='last')
+                  population  GDP alpha-2
+        Anguilla       11300  311      AI
+        Tuvalu         11300   38      TV
+        Nauru         337000  182      NR
+
+        When using ``keep='all'``, the number of element kept can go beyond ``n``
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py:2082]: Is there a missing period here after `n`?

File: src/snowflake/snowpark/modin/plugin/docstrings/series.py [hunk 9]
```diff
@@ -1598,11 +1598,189 @@ def ne():
...
+
+        See Also
+        --------
+        Series.nlargest: Get the `n` largest elements.
+        Series.sort_values: Sort Series by values.
+        Series.head: Return the first `n` rows.
+
+        Notes
+        -----
+        Faster than ``.sort_values().head(n)`` for small `n` relative to
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/docstrings/series.py:1722]: ditto

**Review by sfc-gh-mvashishtha — COMMENTED** [review]
some minor comments

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 10]
```diff
@@ -13948,6 +13948,127 @@ def dt_strftime(self, date_format: str) -> None:
...
+        # Native pandas returns empty dataframe if n is negative. Set it to zero to
+        # provide same behavior.
+        n = max(0, n)
+        # Native pandas returns empty dataframe if 'columns' is a empty array. Set n
+        # to zero to provide same behavior.
+        if len(columns) == 0:
+            n = 0
+
+        # We need row position to break tie for duplicate values.
+        internal_frame = self._modin_frame.ensure_row_position_column()
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:13997]: do you need the row position column? or just the ordering column? If we do not need the row position column, we can just use the current ordering columns

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 11]
```diff
@@ -13948,6 +13948,127 @@ def dt_strftime(self, date_format: str) -> None:
...
+            )
+        )
+        ordering_columns = []
+        for label, identifiers in zip(columns, matched_identifiers):
+            if len(identifiers) == 0:
+                raise KeyError(label)
+            if len(identifiers) > 1:
+                raise ValueError(f"The column label '{label} is not unique.")
+            ordering_columns.append(
+                OrderingColumn(identifiers[0], ascending, na_last=True)
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:14011]: it here any column type that is not sortable in snowflake? for example, variant column etc?

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 10 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py]: good point. we can work with existing ordering columns.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 4 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py]: good catch! added now.

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py (hunk 6 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py]: right it's not. This got copied from pandas doc. Removed now.

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py (hunk 7 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py]: removed.

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py (hunk 8 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py]: Added.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 11 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:14011]: Is suggestion here to add type check?

**Review by sfc-gh-mvashishtha — APPROVED** [review]

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 11 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:14011]: Oh, no. let's do not do special type check, i am just curious about what happens. i think it is important for us to understand potential consequence

File: docs/source/modin/supported/dataframe_supported.rst [hunk 12]
```diff
@@ -279,13 +279,13 @@ Methods
 +-----------------------------+---------------------------------+----------------------------------+----------------------------------------------------+
 | ``ne``                      | P                               | ``level``                        |                                                    |
 +-----------------------------+---------------------------------+----------------------------------+----------------------------------------------------+
-| ``nlargest``                | N                               |                                  |                                                    |
+| ``nlargest``                | P                               |                                  | ``N`` if ``keep == "all"``                         |
 +-----------------------------+---------------------------------+----------------------------------+----------------------------------------------------+
```
Comment (sfc-gh-yzou) [inline docs/source/modin/supported/dataframe_supported.rst:283]: forgot to mention that before, you also need to make sure you update dataframe.rst, series.rst to make sure those two APIs occur in the generated API doc

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 5 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:13981]: Oh, what i was referring to is for example I can call the following 
```
df.nlargest(3, columns=None)
```
pandas raises an keyError. Here I believe you simply want to check if self._modin_frame.is_unnamed_series():
  columns = [MODIN_UNNAMED_SERIES_LABEL]
the logic seems simple and straightforward to understand. And maybe you can also add the above example as a test point

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 13]
```diff
@@ -13948,6 +13948,132 @@ def dt_strftime(self, date_format: str) -> None:
...
+        # Native pandas returns empty dataframe if 'columns' is a empty array. Set n
+        # to zero to provide same behavior.
+        if len(columns) == 0:
+            n = 0
+
+        internal_frame = self._modin_frame
+        # Map pandas labels to snowflake identifiers.
+        matched_identifiers = (
+            internal_frame.get_snowflake_quoted_identifiers_group_by_pandas_labels(
+                columns, include_index=False
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:14001]: want to check that we have test that order by the index columns, and case have duplicated columns among index and regular data columns for example index = ['A', 'B'], columns = ['C', 'B', 'E']?

File: src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py (hunk 6 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/docstrings/dataframe.py]: In fact, depends on how the implementation look like, it might be the case, for example i would expect snowflake top n performs better than order by <> limit n.  however, based on the current implementation, difficult to tell. Safe to do not mention this for now

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 5 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:13981]: Ah didn't know this method existed. Makes sense to use this.

> And maybe you can also add the above example as a test point

We already have a  test case for this in `test_nlargest_nsmallest_invalid_label`.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 11 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:14011]: Are you sure variant is not sortable? Have you verified this?

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 13 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:14001]: TBH given that multi-index is deprioritized. I am not sure if it best use of time to write tests for these edge cases related to multi-index.

File: docs/source/modin/supported/dataframe_supported.rst (hunk 12 above)
Comment (sfc-gh-nkumar) [reply docs/source/modin/supported/dataframe_supported.rst:283]: Done

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 13 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:14084]: i don't think we even need multiIndex for this test, it is just test overlap between index and regular columns. 
Also if we do not intend to support multiIndex here, please make sure we raises an error if multiIndex is involved.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 11 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:14094]: No, I haven't, that is why i am asking here? did we have test for columns like variant, for example which have both string, int, timestamp etc. i think I just want to know what kind of column type we have tried, and have we noticed any behavior difference between Snowpark pandas and Pandas. 
When I look at the test case, it seems we only tried very basic numeric type.

Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:14094]: actually, i tried on a string column for native pandas, it raise TypeError. i don't think we need to have exact the same behavior, but we probably should have a test to indicate such behavior difference.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 13 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:14084]: Will add test to unblock.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 11 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:14094]: I think we need to more precise in code review comments. It's very confusing, first it seems like we aksing for type check then it seems we want to know what happens to unsortable types (which don't exist in snowflake) now we want to tests for string type. What exactly is ask here to unblock this PR?

Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:14094]: I think my ask here is what happens if user is calling nsmallest or nlargest on a column with non-numeric types, which we don't even seem to have a test case for it.  Is there any behavior difference between Snowpark pandas and Pandas? My first though is what happens if the column has a mixed data type, will it error out in Snowflake? then I tried out with just string type, it seems there is definitely a behavior difference between native pandas and Snowpark pandas.
 For example, at least i am sure sort on string type is definitely working in snowflake, but pandas doesn't allow such case, so pandas probably won't allow calling this function on columns with mixed dtype also, or maybe even variant type with the same value type. However, those case may or may not work with snowpark pandas which mainly relies on how snowflake works.

Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:14094]: i think i got the answer here already could you please maybe leave a comment here about such behavior difference between pandas and snowpark pandas, or maybe have a test case to indicate the behavior difference here.

Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:14094]: Added test to unblock.

**Review by sfc-gh-yzou — APPROVED** [review]

File: tests/integ/modin/frame/test_nlargest_nsmallest.py [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+
+
+@sql_count_checker(query_count=1)
+def test_nlargest_nsmallest_large_n(snow_df, native_df, method):
+    eval_snowpark_pandas_result(
+        snow_df, native_df, lambda df: getattr(df, method)(100, "A")
+    )
+
+
+@sql_count_checker(query_count=5, join_count=1)
```
Comment (sfc-gh-helmeleegy) [inline tests/integ/modin/frame/test_nlargest_nsmallest.py:54]: Comment to explain high query count?

**Review by sfc-gh-helmeleegy — APPROVED** [review]
Overall LGTM. Thanks!

File: CHANGELOG.md [hunk 15]
```diff
@@ -49,6 +49,7 @@
 - Added support for `DataFrame.expanding` and `Series.expanding` for aggregations `count`, `sum`, `min`, `max`, `mean`, `std`, and `var` with `axis=0`.
 - Added support for `DataFrame.rolling` and `Series.rolling` for aggregation `count` with `axis=0`.
 - Added support for `Series.str.match`.
+- Added support for `DataFrame.nlargest`, `DataFrame.nsmallest`, `Series.nlargest` and `Series.nsmallest`
```
Comment (sfc-gh-helmeleegy) [inline CHANGELOG.md:52]:
```suggestion
- Added support for `DataFrame.nlargest`, `DataFrame.nsmallest`, `Series.nlargest` and `Series.nsmallest`.
```

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 16]
```diff
@@ -14033,6 +14033,130 @@ def dt_strftime(self, date_format: str) -> None:
...
+            # TODO SNOW-1483214: Add support for keep='all'
+            ErrorMessage.not_implemented(
+                f"Snowpark pandas method '{method_name}' doesn't yet support parameter keep='all'"
+            )
+        # Special case handling for unnamed series. 'columns' passed from frontend layer
+        # will be None, replace it with MODIN_UNNAMED_SERIES_LABEL.
+        if self._modin_frame.is_unnamed_series():
+            columns = MODIN_UNNAMED_SERIES_LABEL
+
+        if not isinstance(columns, list):
```
Comment (devin-petersohn) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:14069]:
```suggestion
        if not is_list_like(columns):
```

**Review by devin-petersohn — COMMENTED** [review]
looks great, one minor issue

**Review by sfc-gh-dpetersohn — APPROVED** [review]
lgtm
